File: go-backend-interview-prep/08_http_and_backend/02_routing_and_handlers/main.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net"
	"net/http"
	"strings"
	"sync"
	"time"
)

// In-memory store (mutex-protected)
var (
	items   = map[string]string{"1": "alpha", "2": "bravo"}
	mu      sync.RWMutex
	nextID  = 3
)

// --- Handlers ---

func listItems(w http.ResponseWriter, r *http.Request) {
	mu.RLock()
	defer mu.RUnlock()
	writeJSON(w, http.StatusOK, items)
}

func getItem(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	mu.RLock()
	defer mu.RUnlock()
	val, ok := items[id]
	if !ok {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "not found"})
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{"id": id, "name": val})
}

func createItem(w http.ResponseWriter, r *http.Request) {
	var body struct{ Name string }
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil || body.Name == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "name required"})
		return
	}
	mu.Lock()
	id := fmt.Sprintf("%d", nextID)
	nextID++
	items[id] = body.Name
	mu.Unlock()
	writeJSON(w, http.StatusCreated, map[string]string{"id": id, "name": body.Name})
}

func deleteItem(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	mu.Lock()
	defer mu.Unlock()
	if _, ok := items[id]; !ok {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "not found"})
		return
	}
	delete(items, id)
	writeJSON(w, http.StatusOK, map[string]string{"deleted": id})
}

// Catch-all: returns 404 JSON for unmatched routes
func notFoundHandler(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusNotFound, map[string]string{
		"error": "route not found",
		"path":  r.URL.Path,
	})
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

// methodCheck wraps a handler to restrict to specific methods (pre-1.22 style)
func methodCheck(method string, next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if !strings.EqualFold(r.Method, method) {
			writeJSON(w, http.StatusMethodNotAllowed, map[string]string{
				"error":   "method not allowed",
				"allowed": method,
			})
			return
		}
		next(w, r)
	}
}

func main() {
	mux := http.NewServeMux()

	// Go 1.22+ style: "METHOD /path"
	mux.HandleFunc("GET /items", listItems)
	mux.HandleFunc("GET /items/{id}", getItem)
	mux.HandleFunc("POST /items", createItem)
	mux.HandleFunc("DELETE /items/{id}", deleteItem)

	// Catch-all for unknown routes
	mux.HandleFunc("/", notFoundHandler)

	// Also show old-style method check for interview reference
	_ = methodCheck // referenced above for pre-1.22 fallback

	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		log.Fatal(err)
	}
	addr := ln.Addr().String()
	fmt.Printf("Server on http://%s\n", addr)
	fmt.Println("  GET    /items       -> list all")
	fmt.Println("  GET    /items/{id}  -> get one")
	fmt.Println("  POST   /items       -> create")
	fmt.Println("  DELETE /items/{id}  -> delete")

	server := &http.Server{Handler: mux}
	go func() {
		time.Sleep(30 * time.Second)
		fmt.Println("\nAuto-shutdown (demo)...")
		server.Close()
	}()

	if err := server.Serve(ln); err != http.ErrServerClosed {
		log.Fatal(err)
	}
}

File: go-backend-interview-prep/08_http_and_backend/02_routing_and_handlers/main.py
"""Routing and handlers -- Python equivalent of the Go example.

Manual path parsing with stdlib http.server (no Flask/FastAPI).
"""

import json
import re
import threading
import urllib.request
from http.server import HTTPServer, BaseHTTPRequestHandler

# In-memory store
items: dict[str, str] = {"1": "alpha", "2": "bravo"}
next_id = 3


def json_response(handler, status: int, data: dict) -> None:
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json")
    handler.end_headers()
    handler.wfile.write(json.dumps(data).encode())


class Router(BaseHTTPRequestHandler):
    """Manual routing -- stdlib has no built-in router."""

    def do_GET(self):
        # GET /items -- list all
        if self.path == "/items":
            json_response(self, 200, items)
            return

        # GET /items/{id} -- get one
        m = re.match(r"^/items/(\w+)$", self.path)
        if m:
            item_id = m.group(1)
            if item_id not in items:
                json_response(self, 404, {"error": "not found"})
                return
            json_response(self, 200, {"id": item_id, "name": items[item_id]})
            return

        json_response(self, 404, {"error": "route not found", "path": self.path})

    def do_POST(self):
        if self.path == "/items":
            global next_id
            length = int(self.headers.get("Content-Length", 0))
            body = json.loads(self.rfile.read(length))
            name = body.get("name", "")
            if not name:
                json_response(self, 400, {"error": "name required"})
                return
            item_id = str(next_id)
            next_id += 1
            items[item_id] = name
            json_response(self, 201, {"id": item_id, "name": name})
            return

        json_response(self, 404, {"error": "route not found"})

    def do_DELETE(self):
        m = re.match(r"^/items/(\w+)$", self.path)
        if m:
            item_id = m.group(1)
            if item_id not in items:
                json_response(self, 404, {"error": "not found"})
                return
            del items[item_id]
            json_response(self, 200, {"deleted": item_id})
            return

        json_response(self, 404, {"error": "route not found"})

    def log_message(self, *args):
        pass


def main() -> None:
    server = HTTPServer(("127.0.0.1", 0), Router)
    host, port = server.server_address
    base = f"http://{host}:{port}"
    print(f"Server on {base}")
    threading.Thread(target=server.serve_forever, daemon=True).start()

    # --- Demo requests ---
    print("\n--- Demo ---")

    # List items
    resp = urllib.request.urlopen(f"{base}/items")
    print(f"GET /items       -> {resp.read().decode()}")

    # Get one
    resp = urllib.request.urlopen(f"{base}/items/1")
    print(f"GET /items/1     -> {resp.read().decode()}")

    # Create
    req = urllib.request.Request(
        f"{base}/items",
        data=json.dumps({"name": "charlie"}).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    resp = urllib.request.urlopen(req)
    print(f"POST /items      -> {resp.read().decode()}")

    # Delete
    req = urllib.request.Request(f"{base}/items/2", method="DELETE")
    resp = urllib.request.urlopen(req)
    print(f"DELETE /items/2   -> {resp.read().decode()}")

    # 404
    try:
        urllib.request.urlopen(f"{base}/nope")
    except urllib.error.HTTPError as e:
        print(f"GET /nope        -> {e.code}: {e.read().decode()}")

    server.shutdown()


if __name__ == "__main__":
    main()
